topdown/http.go

```diff
@@ -568,3 +622,349 @@ func parseTimeout(timeoutVal ast.Value) (time.Duration, error) {
 		return timeout, builtins.NewOperandErr(1, "'timeout' must be one of {string, number} but got %s", ast.TypeName(t))
 	}
 }
+
+func getBoolValFromReqObj(req ast.Object, key string) (bool, error) {
+	result := false
+	var err error
+
+	value := req.Get(ast.StringTerm(key))
+	if value != nil {
+		result, err = strconv.ParseBool(value.String())
+		if err != nil {
+			return false, err
+		}
+	}
+	return result, nil
+}
+
+type persistentCacheValue struct {
+	value        ast.Value         // response
+	size         int               // size of response body
+	date         time.Time         // origination date and time of response
+	cacheControl map[string]string // response cache-control header
+	expires      time.Time         // date/time after which the response is considered stale
+	etag         string            // identifier for a specific version of the response
+	lastModified string            // date and time response was last modified as per origin server
+}
+
+func newPersistentCacheValue(value ast.Value, size int) (*persistentCacheValue, error) {
```

Comment:
There's unnecessary conversions happening here:

* response returned from `net/http` as mix of native Go values (maps, strings, slices, etc.) and struct fields
* response converted in whole into `ast.Value`
* `ast.Value` converted back into native Go values (via `ast.JSON` call)
* native Go values converted back into struct fields on `persistentCacheValue`

I think this is more complex than it needs to be. We should be able to construct the `persistentCacheValue` directly from the response. I think this will simplify things and also make it easier to unit test. E.g., with the current implementation most of the error paths are not covered.

Reply:
Refactored the code to avoid the conversions.